README
DBD::InterBase -- DBI driver for Firebird and InterBase RDBMS server.

   $Id: README,v 1.12 2005/09/12 03:21:41 edpratomo Exp $

   Copyright (c) 1999-2005  Edwin Pratomo
   Portions Copyright (c) 2001-2005  Daniel Ritz 

   You may distribute under the terms of either the GNU General Public
   License or the Artistic License, as specified in the Perl README file.

   PLEASE READ THE ENTIRE README FILE CAREFULLY !

   WARNING: THIS IS BETA SOFTWARE. Your mileage may vary.

REQUIREMENTS:
  - Perl 5.004 installed
  - Perl DBI (1.41 and higher)
  - Firebird or InterBase
  - A C compiler
    * UN*X
      GCC 2.95.2 (tested, older may or may not work) or other capable

    * Windows
      - Borland C++ 5.5
        You can donwload a free copy of BCC55 at
          ftp://ftpd.borland.com/download/bcppbuilder/freecommandLinetools.exe

        In order to compile with Borland compiler you need pmake:
          http://www.cpan.org/authors/id/NI-S/Make-1.00.tar.gz
        dmake would work too (building) but has problems creating PPD file
          http://www.cpan.org/authors/id/GSAR/dmake-4.1pl1-win32.zip


      - or Microsoft VC++ 6.0


*BEFORE* BUILDING, TESTING AND INSTALLING this you will need to:

  - Build, test and install Perl 5 (at least 5.004).
    It is very important to TEST it and INSTALL it!

  - Build, test and install the DBI module (at least DBI 1.41).
    It is very important to TEST it and INSTALL it!

  - Remember to *read* the DBI README file!

  - Make sure that InterBase/Firebird server is running (for testing)


BUILDING:
  Win32 with Borland C++:
    type 'pmake' or 'dmake', not just 'make'. before run make: 
      perl PatchBCC

  Win32 with MS compiler:
    type 'nmake', not just 'make'


  To Configure and build the DBD:
    perl Makefile.PL
      (when asked to specify a test database, please specify a non-existing file)
    make

TESTING

  make test
  make install               (if the tests look okay)



Changes
Revision history for Perl extension DBD::InterBase.

$Id: Changes 381 2007-05-20 15:25:13Z edpratomo $

May 20, 2007 - edpratomo
* fixed retval from ib_set_tx
* removed debug output from ib_tx_info

Oct 25, 2006 - edpratomo
* added ib_tx_info()
* added t/91txinfo.t
* added skip check for t/80event-ithreads.t if Perl version older than 5.8.8
* updated MANIFEST

Oct 24, 2006 - edpratomo
* fixed some croaks in InterBase.xs
* safer t/62timeout.t and t/90dbinfo.t

Oct 23, 2006 - edpratomo
* added support for wait timeout
* added -Wall and some warnings fixes
* dbkey_scope defaults to 0. configurable.
* added active_tran_count and creation_date req items for ib_database_info()
* added tests for ib_database_info() and wait timeout

Oct 14, 2006 - edpratomo
* applied patches from Michael Moehle for 64 bit platform
* added backward compatibility
* ib_error_check() reworked, with fb_interpret()

Sep 12, 2005 - edpratomo
* cleaned warnings when compiled with gcc-4.0

Sep 11, 2005 - edpratomo
* removed unused event state
* renewed dbd_init(); DBIh_EVENT2 call and DBIS were removed.
  perl -pi.old -e
    's/DBI_TRACE\((\d),\s*?\(DBILOGFP, ("[^"]+".*)\)\);/DBI_TRACE_imp_xxh(imp_xyz, $1, (DBIc_LOGPIO(imp_xyz), $2));/' 
     InterBase.xs dbdimp.c
  followed by manual tweaking
* triggers cleanup in t/8[01]event-*.t

Sep 10, 2005 - edpratomo
* Updates to documentation regarding bugs, faq, and event
* fixed $how_many in t/81event-fork.t
* GetInfo.pm was regenerated using easysoft's firebird odbc driver,
  unixodbc, DBI::DBD::Metadata, and DBD::ODBC, of course
* TypeInfo.pm was generated as well, and added to MANIFEST
* t/40alltypes.t, t/40datetime.t, t/41numeric.t now use find_new_table()

Sep 8, 2005 - edpratomo
* FAQ was moved into main documentation, maintaining docs at separate
  places is difficult
* FAQ.pm and Bundle were removed from MANIFEST. Bundle is not necessary.

Sep 4, 2005 - edpratomo
* fixed segfault when destroying event handles in multithreaded app
* increased DBI version in Makefile.PL
* added t/80event-ithreads.t, t/81-event-fork.t, uses Test::More
* updated documentation

Aug 31, 2005 - edpratomo
* cleaned warnings about unused variables, different signedness
* make it compile with old gcc 2.96
* increased required DBI version - 1.41

Aug 30, 2005 - edpratomo
* manually applied patch sf #1042790 by freeduke for dbdimp.c to support
  InterBase 7. typecasting changes are skipped.
* major changes to event API
* manually applied patch sf #1068671 by dstreifert

Aug 8, 2005 - edpratomo
* fixed bug sf #1171702 (memory leak when doing TM type timestamp)

Jan 14, 2005 - danielritz
* better error handling for blobs
* 30insertfetch.t, 40blobs.t with more tests
* prevent a possible buffer overflow in date handling
* don't accept blob bind param for SELECT statement
* remove pointless 'ib_cursorname' attribute. DBD does it automatically

Nov 26, 2004 - danielritz
* add CLONE() method to support threads
* update test scripts for newer DBI

Mar 3, 2004 - edpratomo
* make bind value with blessed scalar work again

Feb 25, 2004 - edpratomo
* Fixed memory write error in dbd_db_login6()
* update doc

Dec 6, 2003 - edpratomo
* Makefile.PL now works with remote test db
* Added $sth->func('ib_plan')
* Updated doc
* $sth->execute() now returns correct row count values for non-select stmt

Nov 19, 2003 - edpratomo
* Fixed bug #844954

Nov 16, 2003 - danielritz
* Make it work with Firebird 1.5

May 20, 2003 - edpratomo
* Fixed problem in ib_init_event() in filling IB_EVENT->names, caused segfault during destruction.
* Added skipping reason in t/60leaks.t

July 25, 2002 - danielritz
* cleanups, speedups

July 18, 2002 - edpratomo
* fixes by Sergey Skvortsov:
  - fix INT64 for FreeBSD
  - fix for ib_role parsing

July 18, 2002 - danielritz
* %lld as standard format for INT64

July 8, 2002 - danielritz
* numeric fixes for negative values like -0.9

July 1, 2002 - danielritz
* trace message can be deactivated

June ?, 2002 - danielritz
* fixes for ib event 
* more cleanups, IB5 fixes

June ?, 2002 - edpratomo
* initial support for ib event

May ?, 2002 - danielritz
* fix EXECUTE PROCEDURE w/o return values
* compile fixes for IB5
* rollback can do isc_rollback_retaining
* cleanups

April 22, 2002 - danielritz
* fix EXECUTE PROCEDURE statments data fetching

April 5, 2002 - edpratomo
* Switching AutoCommit attribute now also affects active softcommit flags.

April 4, 2002 - edpratomo, danielritz
* Added ib_softcommit attribute, isc_commit_retaining now needs to be enabled
  explicitly by users. 

April 4, 2002 - danielritz
* Added set_tx_param() with no param which now resets TPB.

April 4, 2002 - edpratomo
* Updated t/40cursoron.t, t/70nestedon.t to use ib_softcommit
* Makefile.PL code indented properly, now load dbd_dbi_arch_dir() only once, 
  now prompts with directory name, `make clean` cleans trace.txt
* Updated documentation in InterBase.pm and FAQ.pm
* Changed the semantic of -reserving in set_tx_param(), now uses hashref
  instead of arrayref.
* Fixed warnings when compiled with DBI >= 1.20
* Ilya addressed bug #429820 and some bug in sth_ddl.

February 14, 2002 - ilyaverlinsky
* fix DATE, TIME, TIMESTAP problem

February 8, 2002 - ilyaverlinsky
* fix sth_ddl not always being reset

February 2, 2002 - danielritz
* faster  blob write code

January 30, 2002 - danielritz
* updated numeric handing code
* fixed blob code

January 29, 2002
* Fixed wrong value stored through bind param if numeric(*, 0)
* Added t/41numeric.t for exhaustive test on INT64 type.
* Updated version number in FAQ (danielritz)
* Change acceptable oldest perl version in InterBase.pm

January 28, 2002
* Fixed incorrect precision number in INT64 values and now locale-aware.
* changed version number in InterBase.pm
* Applied a patch to type_info_all by Christian Lademann
* Removed trailing precision if INT64 value is an integer.

December 2001, by danielritz
* no longer start a default transaction on init

Makefile.PL

#   $Id: Makefile.PL 369 2006-10-23 20:49:10Z edpratomo $
#
#   Copyright (c) 1999-2005 Edwin Pratomo
#   Portions Copyright (c) 2001-2005 Daniel Ritz
#
#   You may distribute under the terms of either the GNU General Public
#   License or the Artistic License, as specified in the Perl README file,
#   with the exception that it cannot be placed on a CD-ROM or similar media
#   for commercial distribution without the prior approval of the author.

BEGIN { $^W = 0 }       # turn warnings off
BEGIN { require 5.004 } # 5.003 fixes very important bugs

use ExtUtils::MakeMaker 5.16, qw(prompt &WriteMakefile $Verbose);
use Config;
use Carp;
use strict;
use File::Basename;
use vars qw($Registry);


# This DBI must be installed before we can build a DBD.
# For those not using Dynamic loading this means building a
# new static perl in the DBI directory by saying 'make perl'
# and then using _that_ perl to make this one.
use DBI 1.41 ();
use Test::More 0.4;
use DBI::DBD;   # DBD creation tools

my $ib_dir_prefix;

# init stuff
my $IB_Bin_path = '';
my $isql_name;
my @ib_bin_dirs;
my @ib_inc_dirs;
my $ib_lib_dir = '';

################################################################################
# OS specific configuration
################################################################################
if ($Config::Config{osname} eq 'MSWin32')
{
    $isql_name = 'isql.exe';

    # try to find InterBase installation via the registry
    my $ib_bin_dir = '';

    eval
    {
        require Win32::TieRegistry;
        Win32::TieRegistry->import('$Registry');
        $Registry->Delimiter("/");

        my $sw = $Registry->{"LMachine/Software/"} or die "Can't read LMachine/Software key: $^E\n";
        # We have to check more than one keys, because different
        # releases of InterBase have used different key hierarchies.

        my $key = $sw->{"InterBase Corp/InterBase/CurrentVersion/"} ||
              $sw->{"Borland/InterBase/CurrentVersion/"} ||
              $sw->{"Firebird Project/Firebird Server/Instances/"} ||
              $sw->{"FirebirdSQL/Firebird/CurrentVersion/"};

        if (defined($key))
        {
            $ib_bin_dir = $key->{"/ServerDirectory"};
            $ib_bin_dir ||= $key->{"/DefaultInstance"} . "bin";
            $ib_lib_dir = $key->{"/RootDirectory"};
            $ib_lib_dir ||= $key->{"/DefaultInstance"};
        }
    };
    die "Error: $@\n" if $@;

    $ib_lib_dir .= '\\' unless ($ib_lib_dir =~ m|^.*\\$|gi);

    @ib_bin_dirs = ($ib_bin_dir);
    @ib_inc_dirs = ($ib_lib_dir . "SDK\\include", $ib_lib_dir . "include");
}
else
{
    $isql_name = 'isql';
    @ib_bin_dirs = (qw(/usr/interbase/bin /opt/interbase/bin /opt/firebird/bin /usr/bin /usr/local/bin));
    @ib_inc_dirs = (qw(/usr/interbase/include /opt/interbase/include /opt/firebird/include /usr/include));
}

sub locate_dbi_arch_dir {
    my $dbidir = dbd_dbi_dir();
    my @try = map { "$_/auto/DBI" } @INC;
    my @xst = grep { -f "$_/Driver.xst" } @try;
    Carp::croak("Unable to locate Driver.xst in @try") unless @xst;
    Carp::carp( "Multiple copies of Driver.xst found in: @xst") if @xst > 1;
    return $xst[0];
}

################################################################################
# sub test_files - checks if at least one of the files in the list exists
# Paramters:
#  0: directory
#  1: reference to file list
# Return value: true value if at least on file exists, 0 otherwise
################################################################################
sub test_files
{
    my($dir, $files) = @_;
    local $_;
    -f "$dir/$_" && return $_ for @$files;
    0;
}

################################################################################
# sub dir_choice - prompts for a directory
# Parameters:
#  0: prompt string
#  1: reference to directory list
#  2: reference to file list
# Return value: directory name
################################################################################
sub dir_choice
{
    my($prompt, $dirs, $files) = @_;
    my %dirs;
    my $i;
    my $ret;

    test_files($_, $files) && ($dirs{++$i} = $_) for @$dirs;
    for (1..3)
    {
        foreach my $d (sort keys %dirs) {
            my $choice = prompt("$prompt :", $dirs{$d});
            return $choice if test_files($choice, $files);
        }
    }
    print "Cannot proceed. Aborting..\n";
}

################################################################################
# sub make_test_conf - configure for test (save to ./t/test.conf)
# Parameters: <none>
# Return value: <none>
################################################################################
sub make_test_conf
{
    my $test_conf = './t/test.conf';
    my ($dsn, $user, $pass, $path, $db, $host);

    # read cached config if available
    if (-r $test_conf)
    {
        print "\nReading cached test configuration...";
        open F, $test_conf or die "Can't open $test_conf: $!";
        local @ARGV = ($test_conf);
        ($dsn, $user, $pass) = map {chomp;$_} <>;
        ($db) = $dsn =~ /(?:db|database)=([^;]+);/;
        ($host) = $dsn =~ /(?:host|hostname)=([^;]+);/;
        $path = ($host ? "$host:" : '') . $db;
        close F;
    }

    # ask for database path
    DBPATH: {
        for (1..3) {
            last if $path = prompt("\nFull path to your test database: ", $path);
        }
        die "Must specify a test database" unless $path;

        ($db, $host) = reverse split /:/, $path;

        # no longer necessary
        #PFW - isql on windows doesn't seem to work without the localhost in the db path
        #my $hostpath = $path;
        #if ($path =~ /^localhost:(.+)/) {
        #    $hostpath = $1;
        #}

        # if DB doesn't exist ask for creation
        if ((!$host || $host =~ /localhost/) && !-f $db or
            $host && $host !~ /localhost/)
        {
            print <<"EOM";
$path does not exist.
Trying to create the test database..
Please enter a username with CREATE DATABASE permission.
EOM
            $user = prompt("Username :", $user);
            $pass = prompt("Password :", $pass);
            create_test_db($path, $user, $pass);
            last DBPATH;
        }
        else
        {
            print <<"EOM";
$db exists.
Trying to use an existing database..
Please enter a username to connect.
EOM
            $user = prompt("Username :", $user);
            $pass = prompt("Password :", $pass);

            # check dialect
            my $dialect;
            my $gstat = $IB_Bin_path . "/" . test_files($IB_Bin_path, [qw(gstat gstat.exe)]);
            local *GSTAT;
            open GSTAT, "$gstat -user $user -password $pass -h $path |" or die "Can't execute $gstat: $!";
            while (<GSTAT>) { ($dialect) = $_ =~ /dialect\s+(\d)/i and last }

            unless (defined $dialect) {
                print <<"EOM";

Dialect of $path is UNKNOWN.
This test requires a database of dialect 3. You may specify a non-existent database to create a new one.
EOM
                my $is_proceed = prompt("Proceed anyway or create a NEW test database (P/N)?", "P");
                last DBPATH if $is_proceed =~ /P/i;
                $path = undef, goto DBPATH;
            }
            unless ($dialect == 3) {
                print <<"EOM";
The dialect of $path is: $dialect !!.
This test requires a database of dialect 3. Please specify a non-existent database to create a new one.
EOM
                $path = undef, goto DBPATH;
            }
        }
    }

    # save test config to file
    open F, ">$test_conf" or die "Can't write $test_conf: $!";
    print F 'dbi:InterBase:' . ($host ? "host=$host;" : '') . "db=$db;ib_dialect=3;ib_charset=ISO8859_1\n$user\n$pass\n";
    close F;
}

################################################################################
# sub create_test_db - Creates the test database
# Parameters:
#  0: path to testdatabase to be created
#  1: username used to connect to the DB
#  2: password
# Return value: <none>
################################################################################
sub create_test_db
{
    my ($path, $user, $pass) = @_;

    # create the SQL file with CREATE statement
    open(T, ">./t/create.sql") or die "Can't write to t/create.sql";
    while(<DATA>)
    {
        s/__TESTDB__/$path/;
        s/__USER__/$user/;
        s/__PASS__/$pass/;
        print T;
    }
    close T;

    # try to find isql
    my $isql;
    if (-x "$IB_Bin_path/$isql_name") { 
        $isql = "$IB_Bin_path/$isql_name"; 
    } else {
        for (split /:/, $ENV{PATH})
        {
            s#/+$##g;
            if (-x "$_/$isql_name") {
                $isql = "$_/$isql_name"; last; 
            }
        }
    }

    EXEC:
    {
        for (1..3)
        {
            $isql = prompt("Enter full path to isql: ", $isql);
            last EXEC if (-x $isql);
        }
        die "Unable to execute isql. Aborting..";
    }

    #PFW - isql on windows doesn't seem to work without the localhost in the db path
    my $hostpath = $path;
    if ($path =~ /^localhost:(.+)/) {
        $hostpath = $1;
    }
    # if test db directory doesn't exist -> try to create
    my $dir = dirname $hostpath;
    unless (-d $dir)
    {
        print "Can't find $dir. Trying to mkdir..\n";
        system('mkdir', '-p', $dir) == 0
        or die "Can't mkdir -p $dir";
    }

    # try to execute isql and create the test database
    system($isql, '-sql_dialect', 3, '-i', './t/create.sql') == 0
    or die "Fail calling $isql -i t/create/sql: $?";
}

################################################################################
# MAIN
################################################################################

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

# prompt for InterBase bin directory
$IB_Bin_path = dir_choice("InterBase/Firebird bin directory", [@ib_bin_dirs], [qw(gfix gfix.exe)]);

unless(-x $IB_Bin_path)
{
    carp "I cannot find your InterBase/Firebird installation.\nDBD::InterBase cannot build or run without InterBase.\n";
    exit 1;
}

# get InterBase version
my $IBVERSION;
my $GFIX_PATH = $IB_Bin_path . "/" . test_files($IB_Bin_path, [qw(gfix gfix.exe)]);
chop($IBVERSION = `$GFIX_PATH -z 2>&1`);
$IBVERSION =~ s/^gfix version //o;
my $is_final = $IBVERSION =~ /\.6\d+$/ ? 1 : 0;

# prompt for IB include dir
my $ib_inc = dir_choice("InterBase/Firebird include directory", [@ib_inc_dirs], [qw(gds.h ibase.h)]);


# we use a hash for the MakeMaker parameters
my %MakeParams = (
    'NAME'          => 'DBD::InterBase',
    'VERSION_FROM'  => 'InterBase.pm', # finds $VERSION
    'C'             => ['dbdimp.c'],
    'H'             => ['dbdimp.h', 'InterBase.h'],
    'CCFLAGS'       => '-Wall',
    'INC'           => qq(-I"$ib_inc" -I"${\locate_dbi_arch_dir()}"),
    'OBJECT'        => "InterBase.o dbdimp.o",
    'LIBS'          => [''],
    'OPTIMIZE'      => $Config::Config{'optimize'},
    'XSPROTOARG'    => '-noprototypes',
    'dist'          => {COMPRESS=>'gzip -9f', SUFFIX=>'gz'},
    'clean'         => {FILES => "*.xsi *.old t/*.old *~ t/*~ trace.txt t/trace.txt lib/DBD/InterBase/*~ lib/DBD/InterBase/*.old lib/Bundle/DBD/*~ lib/Bundle/DBD/*.old"},
    'realclean'     => {FILES => "t/test.conf"},
);

# the OS specific build environment setup
my $os = $Config::Config{'osname'};
if ($os eq 'MSWin32')
{
    # set up PPM package parameters
    $MakeParams{'AUTHOR'}   = 'Edwin Pratomo (edpratomo@users.sourceforge.net)';
    $MakeParams{'ABSTRACT'} = 'DBD::InterBase is a DBI driver for Firebird and InterBase, written using InterBase C API.';

    my $vc_dir = '';
    if ($Config::Config{'cc'} eq "cl")
    {
        # try to find Microsoft Visual C++ compiler
        eval
        {
            require Win32::TieRegistry;
            Win32::TieRegistry->import('$Registry');
            $Registry->Delimiter("/");

            my $sw = $Registry->{"LMachine/Software/"};



            # We have to check more than one keys, because different
            # releases of Visual C++ have used different key hierarchies.
            my $key = 
                $sw->{"Microsoft/VisualStudio/6.0/Setup/Microsoft Visual C++"} ||
                $sw->{"Microsoft/VisualStudio/7.0/Setup/VC"};

            if (defined($key))
            {
                $vc_dir = $key->{"/ProductDir"};
            }
        };

        my @vc_dirs = ($vc_dir . "/bin");

        my $VC_PATH = dir_choice("Visual C++ directory", [@vc_dirs], [qw(cl.exe)]);

        unless (-x $VC_PATH){
            carp "I can't find your MS VC++ installation.\nDBD::InterBase cannot build.\n";
            exit 1;
        }

        my $vc_inc = $VC_PATH . "/include";
        my $vc_lib = $VC_PATH . "/lib";

        $INC .= " -I\"$vc_inc\"";

        my $ib_lib = dir_choice("InterBase lib directory", 
                            [$ib_lib_dir . "SDK\\lib_ms", $ib_lib_dir . "lib"],
                            [qw(gds32_ms.lib fbclient_ms.lib)]);

        my $cur_libs = $Config::Config{'libs'} ;
        my $cur_lddlflags = $Config::Config{'lddlflags'} ;

        my $lib;
        if (-f "$ib_lib/fbclient_ms.lib")
            { $lib = "$ib_lib/fbclient_ms.lib"; }
        else
            { $lib = "$ib_lib/gds32_ms.lib"; }

        eval "
    sub MY::const_loadlibs {
    '
LDLOADLIBS = \"$lib\" $cur_libs
LDDLFLAGS = /LIBPATH:\"$vc_lib\" $cur_lddlflags
    '
    } ";
    }
    else
    {
        # Borland C++ 5.5
        my $bcc = $Config::Config{'bcc_path'} . "\\";

        my $BCC_PATH = dir_choice("Borland C++ directory",
            [qw(c:/borland/bcc55), $bcc],
            [qw(bin/bcc32.exe)]);

        unless(-x $BCC_PATH)
        {
            carp "I can't find your Borland C++ installation.\nDBD::InterBase cannot build.\n";
            exit 1;
        }

        my $ib_lib = dir_choice("InterBase lib directory", 
                            [$ib_lib_dir . "SDK\\lib", $ib_lib_dir . "lib"],
                            [qw(gds32.lib fbclient.lib fbclient_ms.lib)]);

        my $lib;
        if (-f "$ib_lib/fbclient.lib")
            { $lib = "$ib_lib/fbclient.lib"; }
        elsif (-f "$ib_lib/fbclient_ms.lib")
        {
            # the borland compiler want omf format libs
            system("\"$BCC_PATH\\bin\\coff2omf\"",
                   "$ib_lib/fbclient_ms.lib",
                   "$ib_lib/fbclient.lib");
            $lib = "$ib_lib/fbclient.lib";
        }
        else
            { $lib = "$ib_lib/gds32.lib"; }

        my $bcc_inc = $BCC_PATH . "include";
        $MakeParams{'CCFLAGS'}    = '-a4 -w- -DWIN32 -DNO_STRICT -DNDEBUG -D_CONSOLE -DHAVE_DES_FCRYPT -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DPERL_MSVCRT_READFIX';
        $MakeParams{'OPTIMIZE'}   = "-O2";
        $MakeParams{'INC'}       .= " -I\"$bcc_inc\"";

        eval "
    sub MY::const_loadlibs {
  '
LDLOADLIBS = \$(LIBC) import32.lib $lib cw32.lib
LDDLFLAGS = -L\"$ib_lib\" -L\"$BCC_PATH\\lib\"
  '
  }
    ";
    }
    # Will I need Cygwin rules too?
}
elsif ($os eq 'solaris')
{
    $MakeParams{'LIBS'} = '-lgdsmt -lm -lc';
}
elsif (($os eq 'linux') || ($os eq 'freebsd'))
{
    my $ib_lib = dir_choice("InterBase/Firebird lib directory",
    [qw(/usr/firebird/lib /opt/firebird/lib /usr/interbase/lib /opt/interbase/lib /usr/lib /usr/local/lib)],
    [qw(libgds.a libgds.so libfbclient.so libfbembed.so)]);
    my $lib;

    if (-f "$ib_lib/libfbclient.so")
    {
         my $emb = 0;

         if (-f "$ib_lib/libfbembed.so")
         {
             while ($emb ne 'n' && $emb ne 'y')
                 { $emb = prompt("Build with libfbembed? (y/n)", 'n'); }
         }

         if ($emb eq 'y')
             { $lib = 'fbembed'; }
         else
             { $lib = 'fbclient'; }
    }
    else
         { $lib = 'gds'; }

    my $ldl = ($os eq 'linux') ? '-ldl' : '';

    $MakeParams{'LIBS'} = "-L$ib_lib -l$lib $ldl ";

    # o Commercial IB 5.1.1 for Red Hat Linux 6.x
    #   requires a compatibilty library to resolve
    #   symbols such as _xstat that were removed
    #   from glibc 2.1.  IB 5.6 and later don't need
    #   this, but it doesn't hurt to list the library.
    if (-f ('/lib/libNoVersion-2.1.2.so'))
  {
     $MakeParams{'LIBS'} .= ' -lNoVersion-2.1.2'
  }
}
elsif ($os eq 'hpux')
{
    $MakeParams{'LIBS'} = '-lgds -ldld';
}
elsif ($os eq 'sco')
{
    # Uncomment this line if you use InterBase 4.0 for SCO OSR5:
    # $LIBS = '-b elf -B dynamic -lgds -lsocket -lcrypt_i';

    # Uncomment this line if you use InterBase 5.5 for SCO OSR5:
    $MakeParams{'LIBS'} = '-lgds -lsocket -lcrypt_i -lc -lm';
}
elsif ($os eq 'sunos')
{
    $MakeParams{'LIBS'} = '-lgdslib -ldl';
}
elsif ($os eq 'irix')
{
    $MakeParams{'LIBS'} = '-lgds -lsun';
}
elsif ($os eq 'aix')
{
    $MakeParams{'LIBS'} = '-lgdsshr';
}
elsif ($os eq 'dgux')
{
    $MakeParams{'LIBS'} = '-lgds -lgdsf -ldl -ldgc';
}
elsif ($os eq 'osf1')
{
    $MakeParams{'LIBS'} = '-lgds';
}
elsif ($os eq 'sysv')
{
   $MakeParams{'LIBS'} = '-lgds -lnsl -lsocket -ldl';
}
else
{
    carp "DBD::InterBase is not supported on platform $os.\n";
    exit 1;
}

# create the test config file
make_test_conf();

# and last but not least write the makefile
WriteMakefile(%MakeParams);

sub MY::postamble
{
    return dbd_postamble(@_);
}

package main;

# the data used to create the database creation script
__DATA__
CREATE DATABASE "__TESTDB__" user "__USER__" password "__PASS__";

quit;
